### PR TITLE
Process pipeline richard

### PR DESCRIPTION
**Pull request by richardyz98**
Hi Matt and Sebastian,
This PR contains only the changes relating to the mapping of integers and the change to numpy indexing over loc and iloc. Also contain are some small miscellaneous changes to the documentation and the implementation, such as removing mutation frequencies as a required input from the perform split method and changing the missing data classifier from string/callable to just directly use the callable. I have shelved the Percolation algorithm for now until the next PR.

Thank you in advance for your comments!

### CODE REVIEW

File: test/solver_tests/greedy_variants_test.py [hunk 1]
```diff
@@ -160,13 +148,10 @@ def test_maxcut_base_case_weights_trivial(self):
         mcgsolver = MaxCutGreedySolver(
             character_matrix=cm, missing_char=-1, priors=priors
         )
```
Comment (mattjones315) [inline test/solver_tests/greedy_variants_test.py:150]: I believe it's this test that fails sporadically (i.e. there's some randomization with the weighting). There are two solutions I can see:

1) set a seed for this test so you know you'll get the same behavior each time
2) come up with a more deterministic test case.

File: test/solver_tests/neighborjoining_solver_test.py [hunk 2]
```diff
@@ -243,6 +254,7 @@ def test_pp_solver(self):
 
         self.nj_pp_solver.solve()
         T = self.nj_pp_solver.tree
+        print(T.edges)
```
Comment (mattjones315) [inline test/solver_tests/neighborjoining_solver_test.py:257]: please remove `print` statements from tests.

File: test/solver_tests/spectral_test.py [hunk 3]
```diff
@@ -10,157 +10,149 @@
...
+                list(cm.loc["c1", :]), list(cm.loc["c3", :]), -1, weights
+            ),
+            0,
+        )
+        self.assertEqual(
+            dissimilarity_functions.hamming_similarity_without_missing(
+                list(cm.loc["c2", :]), list(cm.loc["c3", :]), -1, weights
+            ),
+            5,
+        )
```
Comment (mattjones315) [inline test/solver_tests/spectral_test.py:75]: Lots of potential for sporadic behavior due to randomness in the weights, potentially. Take a look at the comments above with suggestions for dealing with this.

**Review by mattjones315 — CHANGES_REQUESTED** [review]
Thanks for putting this together.

Lots of comments here that I think need to be resolved before merging it in. A couple of themes -

(1) There should be more tests written to support the new behavior you're introducing around the ability to specify new transformations for weights, new dissimilarity / similarity calculations, and more. I've left notes here and there for this.

(2) There's some better engineering practices you can take advantage of to reduce duplicated code & clean up some of the default behavior. Some variable names should also be eschewed for clarity (e.g., `w` -> `weights`). 

(3) In some cases, it's unclear if you really want to include transformations (e.g., ILP or for NJ). These transformations will change the properties of the dissimilarity function calculation so in this case. For example, I'm not sure it'd be wise to allow the user to specify `sqrt` as the transformation when we'll be adding similarities. It might be more straightforward for this case for a user to specify a dissimilarity function. At the very least, I don't think we should change the behavior of the `ILPSolver` and the `DistanceSolver` until we get a better handle on this.

I believe that's it for now. Let's handle these points and then I'll get back to reviewing the PR.

File: cassiopeia/solver/GreedySolver.py [hunk 4]
```diff
@@ -163,18 +170,19 @@ def compute_mutation_frequencies(
             pair
 
         """
-        cm = self.unique_character_matrix.loc[samples, :]
+        subset_cm = self.unique_character_matrix[samples, :]
```
[the remark this answers is not in the document]
Comment (richardyz98) [reply cassiopeia/solver/GreedySolver.py:173]: Hey Matt I agree actually. I've updated this functionality, and will update our tests.

File: cassiopeia/solver/missing_data_methods.py [hunk 5]
```diff
@@ -6,12 +6,12 @@
...
+    character_matrix: pd.DataFrame,
     missing_char: int,
-    left_set: List[Union[int, str]],
-    right_set: List[Union[int, str]],
-    missing: List[Union[int, str]],
-) -> Tuple[List[Union[int, str]], List[Union[int, str]]]:
+    left_set: List[int],
+    right_set: List[int],
+    missing: List[int],
+) -> Tuple[List[int], List[int]]:
```
[the remark this answers is not in the document]
Comment (richardyz98) [reply cassiopeia/solver/missing_data_methods.py:14]: Hey Matt,
I looked and it wasn't in the original Cassiopeia, but I've implemented it here!

File: cassiopeia/solver/dissimilarity_functions.py [hunk 6]
```diff
@@ -141,3 +136,79 @@ def hamming_distance(s1: np.array(int), s2: np.array(int)) -> int:
...
+                    d += 2 * w[i][s1[i]]
+                else:
+                    d += 2
+            else:
+                if not w:
+                    d += 1
+        else:
+            if s1[i] != 0 and s2[i] != 0:
+                if w:
+                    d -= w[i][s1[i]] - w[i][s2[i]]
```
[the remark this answers is not in the document]
Comment (richardyz98) [reply cassiopeia/solver/dissimilarity_functions.py:209]: Hi Matt I gave this some deeper thought, and I do think this is the functionality we want here. I will try to explain in standup on 2/2/2021

File: cassiopeia/solver/GreedySolver.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (richardyz98) [reply cassiopeia/solver/GreedySolver.py:173]: Hey Matt on second thought, after thinking it through more deeply, I take this back. I think that calculating the frequencies with duplicates included could induce some weird behavior. First of all, I have a small counterexample that shows that you could easily produce a less parsimonious tree with including duplicates than with excluding duplicates. I think that given that duplicates will be placed as sisters in the reconstructed tree, we should view them as a technical artifact that occurs from divisions at the bottom of tree, and not as informative to which mutations occurred at the top of the tree (in the case of greedy) or as to which triplets need to be resolved by a supertree instance (in the case of maxcut).

[the remark this answers is not in the document]
Comment (richardyz98) [reply cassiopeia/solver/GreedySolver.py:173]: I'll roll back any changes I've made on this front

**Pull request merged**